addons/mail/models/fetchmail.py

```diff
@@ -207,84 +209,87 @@ def _fetch_mails(self):
-                        if num_messages < MAX_POP_MESSAGES or failed_in_loop == num:
+                        if count >= MAX_IMAP_MESSAGES:
                             break
+                elif connection_type == 'pop':
+                    pop_server = server.connect()
+                    (num_messages, total_size) = pop_server.stat()
+                    pop_server.list()
+                    total_remaining += num_messages
+                    for num in range(1, min(MAX_POP_MESSAGES, num_messages) + 1):
+                        (header, messages, octets) = pop_server.retr(num)
+                        message = (b'\n').join(messages)
+                        process_message(message)
+                        pop_server.dele(num)
+                else:
+                    _logger.warning('Unknown server type for %s: %r', server, connection_type)
+            except Exception as e:
+                if raise_exception:
+                    raise ValidationError(_("Couldn't get your emails. Check out the error message below for more info:\n%s", e)) from e
+                else:
+                    _logger.info("General failure when trying to fetch mail from %s server %s.", *server_type_and_name, exc_info=True)
+            finally:
+                if imap_server:
+                    try:
+                        imap_server.close()
+                        imap_server.logout()
+                    except OSError:
+                        _logger.warning('Failed to properly finish %s connection: %s.', *server_type_and_name, exc_info=True)
+                if pop_server:
+                    try:
                         pop_server.quit()
-                except Exception as e:
-                    if raise_exception:
-                        raise ValidationError(_("Couldn't get your emails. Check out the error message below for more info:\n%s", e)) from e
-                    else:
-                        _logger.info("General failure when trying to fetch mail from %s server %s.", server.server_type, server.name, exc_info=True)
-                finally:
-                    if pop_server:
-                        try:
-                            pop_server.quit()
-                        except OSError:
-                            _logger.warning('Failed to properly finish pop connection: %s.', server.name, exc_info=True)
+                    except OSError:
+                        _logger.warning('Failed to properly finish %s connection: %s.', *server_type_and_name, exc_info=True)
+            if count:
+                _logger.info("Fetched %d email(s) on %s server %s; %d succeeded, %d failed.", count, *server_type_and_name, (count - failed), failed)
+                total_done += count
             server.write({'date': fields.Datetime.now()})
+            total_remaining -= 1  # the server was checked
+            self.env['ir.cron']._notify_progress(done=total_done, remaining=total_remaining)
+            cr.commit()
```

Comment:
The logic behind the progress is usually to notify as often as possible (taking into account the commit overhead). In this case we mark the message as seen on the IMAP / POP server, so we probably want to commit at each message.

Reply:
There is already a commit at each message (see `process_message`) which is needed.
I called the progress API after processing a batch of messages for each server.

Optionally, I could add a notify progress inside the function that processes the message.

Reply:
Right, I missed the commit. Indeed, I think that would be more in line with the concept of notifying the progress at each step.

Reply:
I have added the progress inside `process_message`.